Remove adjust_if_recursive

allredj opened the pull request:
Remove unneeded call to `adjust_if_recursive`.

As requested in https://github.com/diffblue/cbmc/pull/4626.
~Based on https://github.com/diffblue/cbmc/pull/4626.~
~Only review last commit.~

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton commented:
Why do we still need the other calls if we're no longer producing array types with clashing sizes?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dd0131ca3a213db6302eb2b2a780e70cd7d390e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111743229

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: decbe68255e33e9b4867649a00b1dee1c67a1f62).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113856857

Commit:
Call substitute_array_access before adding lemma

Prevents conditional types, which may result in failures during
simplification.

Commit:
get_array: Remove adjust_if_recursive

No longer needed as we got rid of all array if exprts before giving the
lemmas to the solver.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3de61a4cda656e209a26b5e0109a232087685100).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117148716
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

majakusber commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -1047,8 +1002,7 @@ static optionalt<exprt> get_array(
 
-  const exprt arr_val =
-    simplify_expr(adjust_if_recursive(super_get(arr), ns), ns);
+  const exprt arr_val = simplify_expr(super_get(arr), ns);
```
How do we know this does the right thing now? Would all string tests start failing otherwise?

allredj replied:
Yes essentially. `simplify_expr` fails an assertion if there are some type mismatches.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3de61a4cda656e209a26b5e0109a232087685100).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117179978

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3de61a4cda656e209a26b5e0109a232087685100).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117264073
